File: zig/basics/03_loops_control_flow.zig
const std = @import("std");
// TOPIC: Loops & Control Flow | zig run 03_loops_control_flow.zig

pub fn main() !void {
    // TODO 1: while loop — with continue expression
    //   var i: usize = 0;
    //   while (i < 10) : (i += 1) {
    //       std.debug.print("{d} ", .{i});
    //   }
    //   // while with optional (null check):
    //   var opt: ?i32 = 42;
    //   while (opt) |val| { std.debug.print("{d}\n", .{val}); opt = null; }

    // TODO 2: for loop over array/slice — with index
    //   const arr = [_]i32{ 10, 20, 30, 40, 50 };
    //   for (arr, 0..) |item, idx| {
    //       std.debug.print("[{d}]={d}\n", .{idx, item});
    //   }
    //   // Multi-array for (must be same length):
    //   for (arr1, arr2) |a, b| { ... }

    // TODO 3: Labeled break with value
    //   const val = blk: {
    //       var x: i32 = 0;
    //       while (x < 100) : (x += 1) {
    //           if (x * x > 50) break :blk x;
    //       }
    //       break :blk -1;
    //   };

    // TODO 4: inline for/while — unrolled at compile time
    //   const fields = @typeInfo(MyStruct).Struct.fields;
    //   inline for (fields) |field| {
    //       std.debug.print("field: {s}\n", .{field.name});
    //   }

    // TODO 5: switch expression
    //   const x: i32 = 5;
    //   const result = switch (x) {
    //       1, 2 => "one or two",
    //       3...7 => "three to seven",
    //       else => "other",
    //   };
    //   // switch must be exhaustive; ranges with ...

    // TODO 6: comptime if — conditional compilation
    //   comptime { if (@sizeOf(usize) == 8) { @compileLog("64-bit platform"); } }
    //   fn platformMessage() []const u8 {
    //       if (comptime @sizeOf(usize) == 8) return "64-bit" else return "32-bit";
    //   }

    // TODO 7: FizzBuzz using switch
    //   var i: u32 = 1;
    //   while (i <= 100) : (i += 1) {
    //       const msg = switch (@intFromBool(i % 15 == 0) * 3 + @intFromBool(i % 3 == 0) * 2 + @intFromBool(i % 5 == 0)) {
    //           3 => "FizzBuzz",
    //           2 => "Fizz",
    //           1 => "Buzz",
    //           else => blk: {
    //               var buf: [10]u8 = undefined;
    //               break :blk try std.fmt.bufPrint(&buf, "{d}", .{i});
    //           },
    //       };
    //       std.debug.print("{s}\n", .{msg});
    //   }

    // TODO 8: defer and errdefer
    //   {
    //       defer std.debug.print("cleanup\n", .{});  // runs at scope exit
    //       // errdefer: only runs on error path
    //       const data = try allocator.alloc(u8, 100);
    //       errdefer allocator.free(data);
    //       try processData(data);  // if this errors, errdefer frees data
    //       allocator.free(data);   // happy path: free normally
    //   }

    std.debug.print("TODO: implement control flow exercises\n", .{});
}

File: c/basics/03_loops_control_flow.c
#include <stdio.h>
// TOPIC: Loops & Control Flow | gcc -o out 03_loops_control_flow.c && ./out

int main(void) {
    // TODO 1: for, while, do-while loops
    //   for (int i = 0; i < 10; i++) printf("%d ", i);
    //   int n = 5; while (n > 0) { printf("%d ", n--); }
    //   do { printf("once at least\n"); } while (0);

    // TODO 2: break and continue
    //   // Print only odd numbers up to 20, stop at first multiple of 15
    //   for (int i = 1; i <= 20; i++) {
    //       if (i % 15 == 0) break;
    //       if (i % 2 == 0) continue;
    //       printf("%d\n", i);
    //   }

    // TODO 3: if / else if / else
    //   int grade = 85;
    //   if (grade >= 90) puts("A");
    //   else if (grade >= 80) puts("B");
    //   else if (grade >= 70) puts("C");
    //   else puts("F");

    // TODO 4: switch statement
    //   char op = '+';
    //   switch (op) {
    //       case '+': printf("add\n"); break;
    //       case '-': printf("sub\n"); break;
    //       default:  printf("unknown\n");
    //   }

    // TODO 5: Ternary operator
    //   int abs_val = x < 0 ? -x : x;

    // TODO 6: goto (valid use: error cleanup in C)
    //   FILE *f = fopen("file.txt", "r");
    //   if (!f) goto cleanup;
    //   // ... do work ...
    //   cleanup:
    //   if (f) fclose(f);

    // TODO 7: Nested loops — print multiplication table 1–10
    //   for (int i = 1; i <= 10; i++) {
    //       for (int j = 1; j <= 10; j++)
    //           printf("%4d", i * j);
    //       putchar('\n');
    //   }

    // TODO 8: Loop with pointer arithmetic
    //   int arr[] = {10, 20, 30, 40, 50};
    //   for (int *p = arr; p < arr + 5; p++) printf("%d\n", *p);

    printf("TODO: implement loop/control-flow exercises\n");
    return 0;
}
